TrafficCapture/coreUtilities/src/main/java/org/opensearch/migrations/tracing/IInstrumentationAttributes.java

```diff
@@ -0,0 +1,109 @@
+package org.opensearch.migrations.tracing;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.metrics.DoubleHistogram;
+import io.opentelemetry.api.metrics.LongCounter;
+import io.opentelemetry.api.metrics.LongHistogram;
+import io.opentelemetry.api.metrics.LongUpDownCounter;
+import io.opentelemetry.api.trace.Span;
+import lombok.NonNull;
+
+import java.time.Duration;
+import java.util.ArrayList;
+
+public interface IInstrumentationAttributes {
+    AttributeKey<Boolean> HAD_EXCEPTION_KEY = AttributeKey.booleanKey("hadException");
+
+    IInstrumentationAttributes getEnclosingScope();
+
+    default Span getCurrentSpan() { return null; }
+
+    default AttributesBuilder fillAttributes(AttributesBuilder builder) {
+        return builder;
+    }
+
+    Throwable getObservedExceptionToIncludeInMetrics();
+
+    default @NonNull Attributes getPopulatedMetricAttributes(AttributesBuilder attributesBuilder) {
+        final var e = getObservedExceptionToIncludeInMetrics();
+        return e == null ? attributesBuilder.build() : attributesBuilder.put(HAD_EXCEPTION_KEY, true).build();
+    }
+
+    default Attributes getPopulatedSpanAttributes() {
+        return getPopulatedSpanAttributes(Attributes.builder());
+    }
+
+    default Attributes getPopulatedSpanAttributes(AttributesBuilder builder) {
+        return getPopulatedSpanAttributesBuilder(builder).build();
+    }
+
+    default AttributesBuilder getPopulatedSpanAttributesBuilder(AttributesBuilder builder) {
+        var currentObj = this;
+        var stack = new ArrayList<IInstrumentationAttributes>();
```

Comment:
Nit: an ArrayDeque may make this a bit easier to follow

Reply:
Good idea.  I'm not sure that the new code is easier to read, but I like it.
```
        var stack = new ArrayDeque<IInstrumentationAttributes>();
        while (currentObj != null) {
            stack.add(currentObj);
            currentObj = currentObj.getEnclosingScope();
        }
        // reverse the order so that the lowest attribute scopes will overwrite the upper ones if there were conflicts
        return StreamSupport.stream(
                Spliterators.spliteratorUnknownSize(stack.descendingIterator(), Spliterator.ORDERED), false)
                .collect(Utils.foldLeft(builder, (b, iia)->iia.fillAttributes(b)));
```

Reply:
I was thinking a more standard `for each`.

Note, the `add` method of a deque is on the front so by the time it's filled it's in the order you want to iterate.

```java
// Fill Attributes in LIFO order to prioritize nested scope's attributes
var attributeAccumulator = Attributes.builder();
for (IInstrumentationAttributes obj : stack) {
  attributeAccumulator = obj.fillAttributes(attributeAccumulator);
}
return attributeAccumulator;
```

If you must use Streams, you can use Streams::reduce or your foldLeft
```java
// Stream directly over the stack in LIFO order to prioritize nested scope's attributes
return stack.stream().collect(
  Utils.foldLeft(Attributes.builder(),
    (accumBuilder, attributes) -> attributes.fillAttributes(accumBuilder))
  );
```